Add a birthday module

CommanderStorm opened the pull request:
other changes:
- added schema_7.sql
- made mensa update every 24 hours instead of 12
- updated requirenments.txt to reflect the move to rythm 😉

CommanderStorm committed:
Merge pull request #1 from timschumi/master

Test

CommanderStorm committed:
Create Schema_7.sql

CommanderStorm committed:
Create Birthday.py

CommanderStorm committed:
Rename Birthday.py to birthday.py

CommanderStorm committed:
Merge pull request #2 from CommanderStorm/CommanderStorm-birthdays

Create Schema_7.sql

CommanderStorm committed:
Rename Schema_7.sql to schema_7.sql

CommanderStorm committed:
Copied this code to another device

CommanderStorm committed:
minor massage refactoring

CommanderStorm committed:
Update schema_7.sql

CommanderStorm committed:
Update birthday.py

CommanderStorm committed:
renamed some things

CommanderStorm committed:
made things more usecase spesific

CommanderStorm committed:
added json style setup

CommanderStorm committed:
added regular expression for recognising dates

CommanderStorm committed:
made mensa.py run every 24 hours

CommanderStorm committed:
fixed typo

CommanderStorm committed:
added correct help messages

CommanderStorm committed:
added aliasses

CommanderStorm committed:
updated database to a more reasonable schema

CommanderStorm committed:
fixed indentation error

CommanderStorm committed:
implemented all the features

CommanderStorm committed:
updated requirements.txt to match current state of affairs

CommanderStorm committed:
cleanup of unneccesary junk

CommanderStorm committed:
somehow I must have removed the formatting on this part..

CommanderStorm committed:
made the bot actually post to the birthdaychannel not just in a random one (uuups..)

CommanderStorm committed:
(uuups..)

CommanderStorm committed:
:monkagiga:

CommanderStorm committed:
updated the module to actually work..

CommanderStorm committed:
... ids like 597134406972342272 dont like to be in int

CommanderStorm committed:
added list command

CommanderStorm committed:
updated BIRTHDAY_CHANNEL_ID to the real deal

CommanderStorm committed:
modified the database pointer to be bound to oone channel

timschumi commented on cogs/birthday.py:
```diff
@@ -0,0 +1,97 @@
+        for user in self.get_user_ids(day, month):
+            text += ":tada: :fireworks: :partying_face: **Alles Gute zum Geburtstag**, <@!{}>" \
+                    " :partying_face: :fireworks: :tada:\n".format(user)
+        await self.bot.fetch_channel(self.BIRTHDAY_CHANNEL_ID).send(text)
```
Let it post wherever the birthday date was originally set up (i.e. save the channel ID in the database)

CommanderStorm replied:
but that would mean, that birthday messages would fly left right and center (metaphorically speaking), 
instead of in just one specific channel, that can be easily silenced

or is there something I am totally not seeing?

CommanderStorm replied:
i added a setup command to move this module to another channel, but I have no idea how to solve this properly..

Reason:
-  this Variable should be fixed -> 
-  must be determined at startup.->
-  a sql querry would be needed.. but since I cant have one static database (to store which channel to write to...)

timschumi commented on cogs/birthday.py:
```diff
@@ -0,0 +1,97 @@
+        await self.bot.fetch_channel(self.BIRTHDAY_CHANNEL_ID).send(text)
+
+    def get_user_ids(self, day: int, month: int) -> [int]:
+        with self.bot.db.get(self.BIRTHDAY_CHANNEL_ID) as db:
```
DBs are per-server, so attaching it to the Birthday channel is wrong here.

timschumi commented on cogs/mensa.py:
```diff
@@ -10,7 +10,7 @@
     def __init__(self, bot):
         self.bot = bot
-        self.bot.register_job(60 * 60 * 12, self.update_entries)
+        self.bot.register_job(60 * 60 * 24, self.update_entries)
```
Should be a seperate patch if any.

timschumi commented on requirements.txt:
```diff
@@ -1,2 +1,2 @@
-discord.py[voice]
+discord
```
While the pip package name does not match the namespace, we should keep it as `discord.py` since that is the actual upstream package.

Just ignore the PyCharm warnings ;-)

timschumi commented on db/schema_7.sql:
```diff
@@ -0,0 +1,8 @@
+CREATE TABLE birthdays
+(
+    userId TEXT NOT NULL UNIQUE,
+    date   INT  NOT NULL,
```
rename to day (or basically anything else than `date`, since that is technically a keyword).

CommanderStorm replied:
in pure sql that may be the case, but not in sqlite.
But yea.. its a bad habbit :D

timschumi commented on cogs/birthday.py:
```diff
@@ -0,0 +1,97 @@
+
+
+class Birthdays(commands.Cog):
+    DATEPATTERN: Pattern[str] = re.compile(r'(((0[1-9])|(1[0-9])|(2[0-9])|(3[0-1]))\.'  # 01.-31.
```
I feel like we could simplify this a lot...

CommanderStorm replied:
I updated the RegEx.. it should be faster now.. no idea how to simplify this elsewise

timschumi commented on cogs/birthday.py:
```diff
@@ -0,0 +1,97 @@
+        self.bot.register_job(60 * 60 * 24, self.congratulate)
+
+    @commands.group(aliases=['birth', 'birthday', 'birthdate', 'geburtstag'])
+    async def birthdays(self, ctx):
```
The `invoked_subcommand` thing isn't needed, there is a better solution: https://github.com/timschumi/TUMbot/blob/master/cogs/reactionroles.py#L12 (note the `invoke_without_command=True`)

timschumi commented on cogs/birthday.py:
```diff
@@ -0,0 +1,97 @@
+                                           r"|"
+                                           r"(((0[1-9])|(1[0-9])|(2[0-9]))\."  # 01.-29.
+                                           r"02\.)")  # february
+    BIRTHDAY_CHANNEL_ID = 666653781366145028
```
We shouldn't hardcode this. Just send the messages wherever things were set up.

timschumi commented on cogs/birthday.py:
```diff
@@ -0,0 +1,97 @@
+    @commands.has_permissions(manage_messages=True)
+    async def list(self, ctx, querry=""):
+        """lists all birthdays (birthdays or userid are possible, jet optional querries)"""
+        with self.bot.db.get(self.BIRTHDAY_CHANNEL_ID) as db:
```
Same here, DBs are bound to servers (and their IDs).

timschumi commented on cogs/birthday.py:
```diff
@@ -0,0 +1,97 @@
+            elif self.DATEPATTERN.match(querry) is not None:  # Birthday as Querry
+                day, month = querry.strip(".").split(".")
+                results = db.execute(
+                    "SELECT userId, date, month FROM birthdays WHERE date = {} AND month = {}".format(day,
```
Please use prepared statements, which are less prone to injections.

CommanderStorm replied:
sorry I totally forgot about those

timschumi commented on cogs/birthday.py:
```diff
@@ -0,0 +1,97 @@
+                await ctx.send(text)
+            else:  # Username as Querry
+                results = db.execute(
+                    "SELECT userId, date, month FROM birthdays WHERE userId LIKE '{}'".format(querry)).fetchall()
```
Same here (prepared statements)

timschumi commented on cogs/birthday.py:
```diff
@@ -0,0 +1,97 @@
+
+    def update_birthday(self, user_id, day, month):
+        with self.bot.db.get(self.BIRTHDAY_CHANNEL_ID) as db:
+            db.execute("INSERT OR REPLACE INTO birthdays (userId, date, month) VALUES ('{}', {}, {})".format(
```
Same here (prepared statements)

timschumi commented on cogs/birthday.py:
```diff
@@ -0,0 +1,97 @@
+    def get_user_ids(self, day: int, month: int) -> [int]:
+        with self.bot.db.get(self.BIRTHDAY_CHANNEL_ID) as db:
+            return db.execute(
+                "SELECT userId FROM birthdays WHERE date = {} AND month = {}".format(day, month)).fetchall()
```
Dito.

timschumi reviewed (changes requested):
Please see the inline comments.

CommanderStorm committed:
Revert "made mensa.py run every 24 hours"

This reverts commit a67f99f7

CommanderStorm committed:
Revert "updated requirements.txt to match current state of affairs"

This reverts commit 4238d8ca

CommanderStorm committed:
renamed date-sqlitecullom to day

CommanderStorm committed:
switched every querry to be parameterised

CommanderStorm committed:
simplified regex

CommanderStorm committed:
invoked_subcommand was removed and replaced with invoke_without_command=True

CommanderStorm committed:
updated from %s/%d parameteratisation to ? parameteratisation

CommanderStorm committed:
added setup command to move the output to another channel

CommanderStorm committed:
moving retains the previous data

timschumi committed:
Revert changes to mensa.py

timschumi committed:
Remove setup() method

timschumi committed:
Resolve remaining issues

- Save target channel for each server
- Loop through servers by connection
- Add a command to change saved target channel
- Simplify query logic

timschumi reviewed (approved):
Looks fine now.
